### PR TITLE
Don't cache so many things

### PR DESCRIPTION
**Pull request by alexweininger**
Purpose of these changes it to reduce the responsibilities of the cache in resource groups down to one. In doing so, it drastically minimizes the amount of changes needed for #426.

Currently:
* Named `ResourceGroupsItemCache`
* Caches the following relationships:
	* `parent item -> child item`
	* `item -> child items`
	* `item -> branch item`
	* `branch item -> item`
* Generated tree item ids

With these changes:
* Renamed to `BranchDataItemCache`
* Stores `item -> branch item`, and `branch item -> item` relationships. That's it.

I accomplished this by altering the internal Resource Groups items, (`SubscriptionItem`, `GroupingItem`, and `ApplicationResourceItem`) to generate their ids on their own. 

And now `GroupingItem` and `ApplicationResourceItem` track their (optional) parent on their own, instead of secretly attaching a parent property.

### CODE REVIEW

File: src/tree/v2/BranchDataItemCache.ts [hunk 1]
```diff
@@ -0,0 +1,25 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { ResourceGroupsItem } from './ResourceGroupsItem';
+
+export class BranchDataItemCache {
+    private readonly branchItemToItemCache: Map<unknown, ResourceGroupsItem> = new Map();
+    private readonly itemToBranchItemCache: Map<ResourceGroupsItem, unknown> = new Map();
```
Comment (bwateratmsft) [inline src/tree/v2/BranchDataItemCache.ts:10]: Where does the `itemToBranchItemCache` get used?

File: src/tree/v2/BranchDataItemCache.ts [hunk 2]
```diff
@@ -0,0 +1,25 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { ResourceGroupsItem } from './ResourceGroupsItem';
+
+export class BranchDataItemCache {
+    private readonly branchItemToItemCache: Map<unknown, ResourceGroupsItem> = new Map();
```
Comment (bwateratmsft) [inline src/tree/v2/BranchDataItemCache.ts:9]: Not critical, but should we change their names to `branchItemToResourceItemCache` / `resourceItemToBranchItemCache`? Or something along those lines?

**Review by bwateratmsft — APPROVED** [review]
Love it. The best code is deleted code :grin:

File: src/tree/v2/BranchDataItemCache.ts (hunk 2 above)
Comment (alexweininger) [reply src/tree/v2/BranchDataItemCache.ts]: Figured we might as well go all the way and name it `branchItemToResourceGroupsItemCache` 😄

**Review by bwateratmsft — APPROVED** [review]

File: src/tree/v2/BranchDataItemCache.ts (hunk 1 above)
Comment (bwateratmsft) [reply src/tree/v2/BranchDataItemCache.ts]: Sounds good. We can always add it back if we find the need to.

**Pull request merged**